tests/Web/Slack/PagerSpec.hs

```diff
@@ -0,0 +1,113 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module Web.Slack.PagerSpec (spec) where
+
+-- base
+import Control.Exception (throwIO)
+import Data.Maybe (fromJust)
+
+-- fakepull
+import Test.Pull.Fake.IO (FakeStream, pull, newFakeStream)
+
+-- hspec
+import Test.Hspec
+
+-- slack-web
+import Web.Slack.Pager
+import Web.Slack.Common
+  ( mkSlackTimestamp
+  , Cursor (..)
+  , Message (..)
+  , MessageType (..)
+  , SlackMessageText (..)
+  )
+import Web.Slack.Conversation
+  ( ConversationId (ConversationId)
+  , mkHistoryReq
+  , mkRepliesReq
+  , HistoryReq (..)
+  , HistoryRsp (..)
+  , RepliesReq (..)
+  , ResponseMetadata (..)
+  )
+import Web.Slack.Types (UserId (..))
+
+-- text
+import qualified Data.Text as Text
+
+-- time
+import Data.Time.Clock (getCurrentTime, nominalDay, addUTCTime)
+
+
+stubbedSendRequest :: FakeStream (Response HistoryRsp) -> a -> IO (Response HistoryRsp)
+stubbedSendRequest stream _request = fromJust <$> pull stream
+
+
+spec :: Spec
+spec = do
+  let prepare = do
+        nowUtc <- getCurrentTime
+        let now = mkSlackTimestamp nowUtc
+            oldest = mkSlackTimestamp $ addUTCTime (nominalDay * negate 20) nowUtc
+            messagesPerPage = 3
+            allResponses :: [Response HistoryRsp]
+            allResponses = do
+              -- According to https://api.slack.com/docs/pagination,
+              -- The last page's cursor can be either an empty string, null, or non-exisitent in the object.
+              (pageN, cursor) <- zip [1..3] ["cursor1=", "cursor2=", ""]
+              let pageNT = Text.pack (show pageN)
+              pure . Right $ HistoryRsp
+                { historyRspMessages = do
+                    messageN <- [1..messagesPerPage]
+                    let messagesPerPageNDT = fromIntegral messagesPerPage
+                        messageNDT = fromIntegral messageN
+                        messageNT = Text.pack (show messageN)
+                        createdBefore = negate $ nominalDay * ((pageN - 1) * messagesPerPageNDT + messageNDT)
+                    pure $ Message
+                        MessageTypeMessage
+                        (Just . UserId $ "U" <> pageNT <> messageNT)
+                        (SlackMessageText $ "message " <> pageNT <> "-" <> messageNT)
+                        (mkSlackTimestamp $ addUTCTime createdBefore nowUtc)
+                , historyRspResponseMetadata = Just . ResponseMetadata . Just $ Cursor cursor
+                }
+        responsesToReturn <- newFakeStream allResponses
+        return (now, oldest, messagesPerPage, allResponses, responsesToReturn)
+
+  describe "conversationsHistoryAllBy" $
+    it "collect all results by sending requests" $ do
+      (now, oldest, messagesPerPage, allResponses, responsesToReturn) <- prepare
```

Comment:
Since `prepare` is called for all tests here, would it benefit from using [`before`](https://hackage.haskell.org/package/hspec-2.7.4/docs/Test-Hspec.html#v:before) or [`beforeAll`](https://hackage.haskell.org/package/hspec-2.7.4/docs/Test-Hspec.html#v:beforeAll)?

Reply:
I personally don't like `before` etc. because functions like `prepare` here is easier to understand how it works.
Current Hspec's architecture is not suitable for RSpec's `before` etc.

Reply:
Sounds good.